mirrord/layer/src/debugger_ports.rs

```diff
@@ -0,0 +1,158 @@
+                        _ => None,
+                    })
+                } else {
+                    None
+                }
+            }
+        }
+        .and_then(|addr| match addr.ip() {
+            IpAddr::V4(Ipv4Addr::LOCALHOST) | IpAddr::V6(Ipv6Addr::LOCALHOST) => Some(addr.port()),
+            other => {
+                warn!(
+                    "Debugger uses a remote socket address {:?}! This case is not yet handled properly.",
+                    other,
+                );
+                None
+            }
+        })
+    }
+}
+
+/// Local ports used by the debugger running the process.
+/// These should be ignored by the layer.
+#[derive(Debug)]
+pub struct DebuggerPorts {
+    detected: Option<u16>,
+    fixed: Option<RangeInclusive<u16>>,
+}
+
+impl DebuggerPorts {
+    /// Create a new instance of this struct based on the environment variables
+    /// ([`MIRRORD_DETECT_DEBUGGER_PORT_ENV`] and [`MIRRORD_IGNORE_DEBUGGER_PORTS_ENV`]) and command
+    /// line arguments.
+    ///
+    /// Log errors (like malformed env variables) but do not panic.
+    pub fn from_env() -> Self {
+        let detected = env::var(MIRRORD_DETECT_DEBUGGER_PORT_ENV)
+            .ok()
+            .and_then(|s| {
+                DebuggerType::from_str(&s)
+                    .inspect_err(|e| {
+                        error!(
+                            "Failed to decode debugger type from {} env variable: {}",
+                            MIRRORD_DETECT_DEBUGGER_PORT_ENV, e
+                        )
+                    })
+                    .ok()
+            })
+            .and_then(|d| d.get_port(&std::env::args().collect::<Vec<_>>()));
+
+        let fixed = env::var(MIRRORD_IGNORE_DEBUGGER_PORTS_ENV).ok().and_then(|s| {
+            let chunks = s.split('-').map(u16::from_str).collect::<Result<Vec<_>, _>>().inspect_err(|e| error!("Failed to decode debugger ports from {} env variable: {}", MIRRORD_IGNORE_DEBUGGER_PORTS_ENV, e)).ok()?;
+            match *chunks.as_slice() {
+                [p] => Some(p..=p),
+                [p1, p2] if p1 <= p2 => Some(p1..=p2),
+                _ => {
+                    error!(
+                        "Failed to decode debugger ports from {} env variable: expected a port or a range of ports",
+                        MIRRORD_IGNORE_DEBUGGER_PORTS_ENV,
+                    );
+                    None
+                },
+            }
+        });
+
+        Self { detected, fixed }
+    }
+
+    /// Return whether the given [SocketAddr] is used by the debugger.
+    pub fn contains(&self, addr: &SocketAddr) -> bool {
+        matches!(
+            addr.ip(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST) | IpAddr::V6(Ipv6Addr::LOCALHOST)
+        ) && (self.detected == Some(addr.port())
+            || self
+                .fixed
+                .as_ref()
+                .map(|r| r.contains(&addr.port()))
+                .unwrap_or(false))
```

Comment:
There is only one debugger where this could happen with anyways, so I would say just make sure it's safe to do it with that debugger and make the change.

Reply:
Cool! I made the change.
I left setting the ignored range in the extension though, so it works as before for other languages. Looks like debugging Node and plain C still works when I remove it, but I have no idea why. Debugger needs some sort of communication channel with the IDE, does it?